docs/pii-encryption.md

```diff
@@ -0,0 +1,125 @@
+
+## Design considerations
+
+Almost everyone will agree that sensitive information should be encrypted.
+However, merely saying "data is encrypted" is insufficient in real-world
+systems; instead we have to be concerned with implementation specifics,
+including key management.
+
+Many systems encrypt data at rest with a single online key.
+Unfortunately, many attacks which compromise the
+ciphertext will also compromise the key, so these designs provide limited
+protection.
+
+We delineate two kinds of PII: sensitive and non-sensitive. Non-sensitive PII
+consists of email address and phone number, both used for multi-factor authentication.
+Sensitve PII includes everything else a verified user might register with us:
+full name, address, date of birth, and SSN.
+
+Therefore, we have adopted the following design goals:
+
+* Encryption should rely on a multi-factor model, just like authentication does.
+* Long term sensitive PII storage should be encrypted under a key derived from the user's
+  password.
+* Short term (session length) sensitive PII storage may be encrypted
+  under a key controlled exclusively by the server.
+* Sensitive PII only needs to be available during the user session. Offline/background
+  access is not necessary. The exception is non-sensitive PII (email and phone).
+* Forgetting your password should not necessarily result in all PII being lost.
+
+## Implementation
+
+To address all these design goals we have implemented a multi-factor
+encryption model.
+
+When a user first creates an account, we use the user's password,
+a server-controlled random string, and a hardware security module (HSM)
+to create an encryption key.
+Any time the user gives us sensitive personal information
+we use that encryption key to encrypt the personal information.
+The sensitive PII may only be decrypted if all three factors are present: the HSM,
+the user's password, and the random string.
+
+Based on consultation with the [National Institute of Standards and Technology](https://www.nist.gov/)
+we follow these steps to create the encryption key:
+
+* generate and store a random, 160-bit string as `salt`
+* using the password and the `salt`, create a [SCrypt](https://github.com/pbhogan/scrypt) hash
+* split the SCrypt hash into two 128-bit strings, `Z1` and `Z2`
+* generate a random 256-bit string `R`
+* using the HSM, encrypt `R` as `encrypted_R`
+* calculate the [XOR](https://en.wikipedia.org/wiki/Exclusive_or)
+  of `encrypted_R` and `Z1` as `D`
+* store `D` as the encryption key
+* create a SHA-256 hash of the concatenation of `Z2` and `R` called `E`
+* create a SHA-256 hash of `E` called `F`
+* store `F` as the password
+* using the key `E`, encrypt the PII using AES GCM 256-bit encryption
+  as `C`
+* store `C` as the encrypted PII
+
+You can review [the tests for this model in our public repository](https://github.com/18F/identity-idp/blob/master/spec/services/pii/nist_encryption_spec.rb).
```

Comment:
since this is in the same repo, we can just make this a relative URL `../spec/services/pii/nist_encryption_spec.rb` and Github's magical linking will do the right thing